Add tests of eglot--guess-contact, including a failing one, then fix it

dandavison opened the pull request:
See #447. This PR adds unit test coverage for `eglot--guess-contact`. I believe that there is at least one bug in master, which these tests reveal: the `(host-string port-int)` version fails with `(wrong-type-argument stringp ...)`.

@joaotavora I have time to work on this if it's helpful, but would you be able to help me nail down the test assertions here? I've left some TODOs in the test code and will comment inline to make the discussion as concrete as possible. (Or of course feel free to complete the test cases added here yourself!)

dandavison commented on eglot-tests.el:
```diff
@@ -847,6 +847,67 @@ pyls prefers autopep over yafp, despite its README stating the contrary."
+                 ("valid-executable"))))
+
+;; TODO: test invalid executable
+
```
```emacs-lisp
("valid-executable")
```
This test case passes, as expected. However, currently `("invalid-executable")` would pass also. Should it error?

joaotavora replied:
> However, currently ("invalid-executable") would pass also. Should it error?

Depends, what is an "invalid" executable for you?

joaotavora replied:
If you mean an executable that doesn't exit yet, I think it shouldn't be `eglot--guess-contact` job to check this.  It's eventually reach `make-process` in the same stack call and _that_ will check.

dandavison replied:
Right. I've removed all assertions regarding executable validity.

joaotavora commented on eglot-tests.el:
```diff
@@ -847,6 +847,67 @@ pyls prefers autopep over yafp, despite its README stating the contrary."
         (((CodeAction) _title _edit _command)
          (ert-fail "Shouldn't have destructured this object as a CodeAction")))))))
 
+(defun eglot--guess-contact--with-patches (contact-spec)
```
I know there is another symbol with this naming convention in the file, but please don't use it.  Stick with `eglot--foo-bar` for internal symbols.  Also "with patches" ins't very obvious, is it? Nname it something like `eglot--mock-guess-contact`, and give it a quick docstring.

joaotavora replied:
Right, so I know I proposed `eglot--mock-guess-contact`, and you did and gave it a nice long docstring.  Sorry to have you made waste some effort there, but it seems like it's mostly unneeded and can be almost be replaced.  A test for this could look like:
```
(ert-deftest decipher-network-specs ()
  (let ((eglot-server-programs '((foo-mode ...)))
        (major-mode 'foo-mode) (buffer-file-name "bla.foo"))
    (eglot--guessing-contact (guessed-mode guessed-class guessed-contact)
      ;; ... assertions with guessed-mode, guessed-class and guessed-contact
      ;; though guessed-mode might be a bit silly
```

That's easier to read because it relies on a known external interface right away: the semantics of which are a big part of what you're testing.

dandavison replied:
No worries! Yes I agree that's better. I'll switch to that, and if I get time, see if I can add the missing test cases and the fix.

dandavison commented on eglot-tests.el:
```diff
@@ -847,6 +847,67 @@ pyls prefers autopep over yafp, despite its README stating the contrary."
+  (eglot--do-guess-contact-test
+   '("somehost.example.com" 7777 :type network)
+   `(,major-mode dummy-project eglot-lsp-server
+                 ("somehost.example.com" 7777 :type network))))
```
```emacs-lisp
("somehost.example.com" 7777)
("somehost.example.com" 7777 :type network)
```

These both fail but I think they should pass. That's #447

joaotavora replied:
Whas is "pass" for you?  I can't  macroexpand the `eglot--do-guess-contact-test` in my mind right now.  Make two named `ert-deftest`s, then show the ones that fail and shouldn't (and, who knows, the ones that pass and shouldn't).

joaotavora commented on eglot-tests.el:
```diff
@@ -847,6 +847,67 @@ pyls prefers autopep over yafp, despite its README stating the contrary."
+            (buffer-file-name "fake-buffer-file-name"))
+        (eglot--guess-contact))))
+
+(defmacro eglot--do-guess-contact-test (contact-spec expected-contact-data)
```
It'd be slightly more readable in the tests themselves if this was changed to a  `eglot--with...` macro.  Meaning a macro that does exactly the same but binds `mode`, `class`, `contact` so they can be used in the `body`.  This would be the signature:
```
(cl-defmacro eglot--with-contact-test (contact-spec (guessed-name-sym guessed-class-sym guessed-contact-sym) &body body))
```

There would be a small about of repetition of the `should` forms, which normally isn't good.  But in tests, this kind of repetition _is_ good.  In my opinion, of course.

dandavison commented on eglot-tests.el:
```diff
@@ -847,6 +847,67 @@ pyls prefers autopep over yafp, despite its README stating the contrary."
+                 ("somehost.example.com" 7777 :type network))))
+
+;; TODO: test (PROGRAM [ARGS...] :autoport [MOREARGS...])
+;; TODO: test (CLASS-NAME . INITARGS)
```
I haven't added these test cases yet because I'm new to eglot and didn't want to accidentally enshrine an incorrect spec in the tests

joaotavora replied:
That's OK: you can enshrine, we'll de-enshrine if needed before merging to master :-)

dandavison commented:
Excellent, thanks very much for the review. So let's take this one step at a time. With the commits I just pushed, have I got the test helper macro in the form you had in mind?

joaotavora commented:
I added a comment or two there.  There's more that needs to be reviewed here, but I don't have time right now.  I'll try to get to this tomorrow or during next week.

dandavison commented:
I've added one more test case, and squashed the commits. I think that we may have coverage for the spec outlined in the docstring of `eglot-server-programs`. Thanks very much for the lisp instruction by the way!

I haven't yet added a minimal fix for the two failing test cases.

Hopefully by adding tests I've earned the right to speculate about refactoring :) I'm not saying it's the most important area of development in the project, but I wonder about ultimately refactoring the function along these lines:

```emacs-lisp
(defun eglot--guess-contact
    (or (eglot--guess-contact-without-user-input)
        (eglot--resolve-contact-with-user-input)))
```

joaotavora commented:
Yes two things are missing:

-  fix the existing cases.
- minor renames of new the tests themselves.  Let's prefix-them them "eglot-server-programs" for lack of a better convention. And no `--` is needed I think.  They should just be `(ert-deftest server-eglot-server-programs-simple-executable...)` , `(ert-deftest eglot-server-programs-function-with-custom-class...)` , etc, etc

> Hopefully by adding tests I've earned the right to speculate about refactoring :) 

very much so, yes :-)

Well after you fix this you can program "against the rails". If the code becomes simpler (where simpler here is kinda synonymous with "less lines"), I'm all for it.

dandavison commented:
OK, renamed the tests, and I believe 0b879dc70ac49b09132146d0a71f9661e16c253b is the best minimal fix for the bug in the host-and-port case.

dandavison commented:
Outside the tests, I've confirmed that I can now connect using

```emacs-lisp
(setf (alist-get 'python-mode eglot-server-programs) '("localhost" 7777))
```

whereas on master that results in `Wrong type argument: stringp, 7777`.

dandavison commented:
Do you need me to reformat the commit messages in Emacs changelog style?

I had valid FSF copyright papers 10 years ago. Do I need to investigate the status there (if so do you know what my next step would be)?

(I might not get to any refactoring for a while so let's not block on that.)

joaotavora commented:
> Do you need me to reformat the commit messages in Emacs changelog style?

Yes please.

> I had valid FSF copyright papers 10 years ago. Do I need to investigate the status there (if so do you know what my next step would be)?

I don't think they expire.

dandavison commented:
Alright, rebased on master and commit messages reformatted.
